libraries/lib-music-information-retrieval/MusicInformationRetrieval.cpp

```diff
@@ -169,4 +145,87 @@ std::optional<MusicalMeter> GetMusicalMeterFromSignal(
    return GetMeterUsingTatumQuantizationFit(
       decimatedAudio, tolerance, progressCallback, debugOutput);
 }
+
+void SynchronizeProject(
+   const std::vector<std::shared_ptr<AnalyzedAudioClip>>& clips,
+   ProjectInterface& project)
+{
+   const auto isBeatsAndMeasures = project.ViewIsBeatsAndMeasures();
+   const auto importedOnEmptyProject = project.ImportedOnEmptyProject();
+
+   if (!importedOnEmptyProject && !isBeatsAndMeasures)
+      return;
+
+   const auto projectTempo = !importedOnEmptyProject ?
+                                std::make_optional(project.GetTempo()) :
+                                std::nullopt;
+
+   if (!std::any_of(
+          clips.begin(), clips.end(),
+          [](const std::shared_ptr<AnalyzedAudioClip>& clip) {
+             return clip->GetSyncInfo().has_value();
+          }))
+      return;
+
+   Finally Do = [&] {
+      // Re-evaluate if we are in B&M view - we might have convinced the user to
+      // switch:
+      if (!project.ViewIsBeatsAndMeasures())
+         return;
+      std::for_each(
+         clips.begin(), clips.end(),
+         [&](const std::shared_ptr<AnalyzedAudioClip>& clip) {
+            clip->Synchronize();
+         });
+      project.ClipsWereSynchronized();
```

Comment:
I don't quite understand. Could you pseudo-code what you mean?

Reply:
```
//Simplify interface by removing some of virtual getters
class ProjectInterface {
   virtual ~SynchronizeClipsDelegate() = default;
   virtual void OnTimeSignatureRecognized(...) = 0;
   virtual MusicFileImportPrefs GetUserPrefs(...) const = 0;
   virtual void OnClipsSyncronizeComplete() = 0;
};
...
// New function signature:
// doSynchronize replaces ProjectInterface::ViewIsBeatsAndMeasures
// projectTempo replaces ProjectInterface::GetTempo(), but it seem to be not used at all
void SynchronizeProject(
   const std::vector<std::shared_ptr<AnalyzedAudioClip>>& clips,
   ProjectInterface& project, bool projectWasEmpty, bool doSynchronizeClips, std::optional<double> projectTempo)
```
It also looks like that there is an invariant `projectWasEmpty == !projectTempo.has_value()`, so one of these arguments may be reduntant

Reply:
I see.
There's one subtlety: we must evaluate `project.ViewIsBeatsAndMeasures()` after the prompt. We don't know what view the project will be in when this function exits.
And since we need the `ProjectInterface` for that, we might as well use it to get the project tempo, too.